llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -3183,6 +3183,18 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
       return new ICmpInst(Pred, X, ConstantInt::get(Ty, NewC));
   }
 
+  CmpInst::Predicate ChosenPred = Pred.getPreferredSignedPredicate();
+
+  if (Pred.hasSameSign() && Add->hasNoSignedWrap() &&
```

Comment:
```suggestion
  if (Add->hasNoSignedWrap() &&
```
Also remove the signed pred path in the if statement above.
```
(Add->hasNoSignedWrap() &&
       (Pred == ICmpInst::ICMP_SGT || Pred == ICmpInst::ICMP_SLT)) ||
      (Add->hasNoUnsignedWrap() &&
       (Pred == ICmpInst::ICMP_UGT || Pred == ICmpInst::ICMP_ULT))
```
->
```
(Add->hasNoUnsignedWrap() &&
       (Pred == ICmpInst::ICMP_UGT || Pred == ICmpInst::ICMP_ULT))
```
```
APInt NewC =
        Cmp.isSigned() ? C.ssub_ov(*C2, Overflow) : C.usub_ov(*C2, Overflow);
```
->
```
APInt NewC = C.usub_ov(*C2, Overflow);
```

Reply:
It seems to me that the `samesign` flag is required for the transformation to be correct. See [this](https://alive2.llvm.org/ce/z/WzjQ5M) Alive2 link for a counterexample.

Since the `samesign` seems to be required, I don't know if this modification should be done.

Reply:
You're right. I'll make the change.